File: 2024/main.ts
import { parseArgs } from "jsr:@std/cli/parse-args";

const d1GetLists = (data: string): [number[], number[]] => {
  const [l, r]: [number[], number[]] = [[], []];

  data.split("\n").map((line) => {
    const [ln, rn] = line.split("   ");
    l.push(Number(ln));
    r.push(Number(rn));
  });

  return [l, r];
};

function d1p1(data: string): number {
  const [left, right] = d1GetLists(data);
  const [ls, rs] = [left.toSorted(), right.toSorted()];
  return ls.map((l, i) => Math.abs(l - rs[i])).reduce((acc, n) => acc + n, 0);
}

const makeCounter = (data: number[]): Map<number, number> => {
  const counter = new Map();

  for (const n of data) {
    counter.set(n, (counter.get(n) || 0) + 1);
  }

  return counter;
};

function d1p2(data: string): number {
  const [left, right] = d1GetLists(data);
  const rightCounts = makeCounter(right);
  return left
    .map((l) => l * (rightCounts.get(l) || 0))
    .reduce((acc, n) => acc + n, 0);
}

type ProblemName = `d${1 | 2}p${1 | 2}`;

type Solution = (data: string) => number;

const Registry: Map<ProblemName, Solution> = new Map([
  ["d1p1", d1p1],
  ["d1p2", d1p2],
]);

async function loadData(problem: string): Promise<string> {
  const decoder = new TextDecoder("utf-8");
  return decoder
    .decode(await Deno.readFile(`./2024/inputs/${problem}.txt`))
    .trim();
}

if (import.meta.main) {
  const flags = parseArgs(Deno.args, {
    default: { day: 1, part: 1 },
  });
  const problem: ProblemName = `d${flags.day}p${flags.part}`;

  console.log(Registry.get(problem)!(await loadData(problem)));
}

File: 2024/bin/main.ml
let re = Str.regexp "   "

let split_line (line: string): string list = Str.split re line

let read_data (file: string) : string =
  let in_chan = open_in file in
  try
    let contents = really_input_string in_chan (in_channel_length in_chan) in
    close_in in_chan;
    String.trim contents
  with e ->
    close_in_noerr in_chan; (* Still close channel in case error *)
    raise e

let d1_parse(data: string): int list * int list = 
  String.split_on_char '\n' data
  |> List.map split_line
  |> List.fold_left (
    fun (left, right) sublist ->
      match sublist with
      | [x; y] -> (int_of_string x :: left, int_of_string y :: right)
      | _ -> failwith "incorrect sublist length")
    ([], []) 
  |> fun (left, right) ->
    (List.sort compare left, List.sort compare right)

let d1p1 (data: string): int =
  d1_parse data
  |> fun (left, right) ->
    left |> List.mapi (fun i x -> (Int.abs(x - List.nth right i)))
  |> List.fold_left ( + ) 0

let () = 
  let day = ref 1 in
  let part = ref 1 in

  let speclist = [
    ("--day", Arg.Set_int day, "AoC Day");
    ("--part", Arg.Set_int part, "AoC Part");
  ] in

  Arg.parse speclist (fun _ -> ()) "Usage: aoc [--day day] [--part part]";

  Printf.sprintf "./inputs/d%dp%d.txt" !day !part
  |> read_data
  |> match !day, !part with
    | 1, 1 -> d1p1
    | _ -> failwith "unknown day or part";
  |> string_of_int
  |> print_endline
